### PR TITLE
Add option to scale number of replicas to zero

### PR DESCRIPTION
**Pull request by atantawi**
- This PR is about adding support to the WVA to scale down the number of replicas of a variant to zero, as opposed to the current minimum of one.
- Scaling to zero occurs when there is no request traffic to a variant.
- A configuration boolean environment variable `WVA_SCALE_TO_ZERO` is added to the `variantautoscaling-config` map, with a default value `true`.
- **The default behavior is scale to zero enabled**. If not desired, `WVA_SCALE_TO_ZERO` has to be set to `false`.
- Scaling up from zero is not addressed by this PR.

### CODE REVIEW

**Comment by atantawi** [comment]
Configuration boolean option `SCALE_TO_ZERO` is added.
The default is `true`.
Option is set in the `variantautoscaling-config` ConfigMap.

**Review by asm582 — CHANGES_REQUESTED** [review]
initial review

**Review by asm582 — CHANGES_REQUESTED** [review]
env var name suggestion.

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -80,6 +80,11 @@ spec:
                 key: PROMETHEUS_TLS_INSECURE_SKIP_VERIFY
           - name: PROMETHEUS_TOKEN_PATH
             value: "/var/run/secrets/kubernetes.io/serviceaccount/token"
+          - name: SCALE_TO_ZERO
```
[the remark this answers is not in the document]
Comment (atantawi) [reply config/manager/manager.yaml]: name changed to `WVA_SCALE_TO_ZERO`

**Comment by atantawi** [comment]
@WheelyMcBones This PR makes scale to zero as default. Accordingly, I changed the desired number of replicas to zero in the no load e2e test case. Do we need to add a test for the no load, scale to zero disabled case?

**Comment by atantawi** [comment]
> @atantawi @asm582 in `actuator.go`:
> 
> ```go
> func (a *Actuator) getCurrentDeploymentReplicas(ctx context.Context, va *llmdOptv1alpha1.VariantAutoscaling) (int32, error) {
> 	var deploy appsv1.Deployment
> 	err := utils.GetDeploymentWithBackoff(ctx, a.Client, va.Name, va.Namespace, &deploy)
> 	if err != nil {
> 		return 0, fmt.Errorf("failed to get Deployment %s/%s: %w", va.Namespace, va.Name, err)
> 	}
> 	// Prefer status replicas (actual current state)
> 	if deploy.Status.Replicas > 0 {
> 		return deploy.Status.Replicas, nil
> 	}
> 	// Fallback to spec if status not ready
> 	if deploy.Spec.Replicas != nil {
> 		return *deploy.Spec.Replicas, nil
> 	}
> 	// Final fallback
> 	return 1, nil
> }
> ```
> 
> This function will never return the `deploy.Status.Replicas` if the deployment is scaled to zero replicas, right?

Fixed

**Comment by WheelyMcBones** [comment]
> * Scaling up from zero is not addressed by this PR.

Possibly missing something here, but once a VA is applied, if no load is currently being sent to the Deployment, Inferno will immediately suggest a scale down to 0.
If scaling up from zero is not there, then we will be stuck with zero replicas?

**Comment by atantawi** [comment]
The point is that once a VA is scaled down to zero, the deployment and the VA object will be deleted. The process of recreating those objects, once traffic to the VA model returns, is out of scope of this PR.

**Comment by asm582** [comment]
> The point is that once a VA is scaled down to zero, the deployment and the VA object will be deleted. The process of recreating those objects, once traffic to the VA model returns, is out of scope of this PR.

The deployment should not be deleted once the VA is scaled to zero. 
I assume that, based on the traffic rate and SLOs being violated, the optimizer will choose an earlier scaled zero replica to be scaled up to N. Is that correct @atantawi ?
